### DISCUSSION TITLE
How can I use Microsoft.Maui.Graphics in UNO?

### DISCUSSION THREAD

**Discussion by lindexi** [unoplatform/uno#15097] [Q&A] [answered]
My fundamental issue is that I am preparing to develop an application with writing functionality for use on touchscreen devices, but I find that [InkCanvas is missing](https://github.com/unoplatform/uno/issues/1821 ). Given my extensive experience in touchscreen and handwriting algorithm development, I aim to add an InkCanvas control filled with custom logic on UNO. However, I find that there is a lack of a basic mechanism in UNO, a mechanism that allows for arbitrary drawing on the interface, similar to the capabilities provided by Microsoft.Maui.Graphics or the capabilities of WPF's DrawingContext. Currently, I can only indirectly achieve this function through Microsoft.UI.Xaml.Shapes, but the controllability and performance of this method are poor. I have written a simple [UnoInk](https://github.com/lindexi/lindexi_gd/tree/8d59a96e0d4e390ae78946ff556a759901961856/HallgaiwhiyiwaLejucona/UnoInk) test application for this, and after three months of development, I find it difficult to implement a feature that can be used in productization without the help of features like Microsoft.Maui.Graphics or WPF's DrawingContext and it is still far from production-ready.

In fact, I believe that if I could interface with Microsoft.Maui.Graphics, or even just open up Skia's drawing capabilities to UIElement, etc., I would be capable of creating a fairly good interface. As I am relatively new to UNO, if there are already similar features implemented in UNO, please let me know.

I look forward to your assistance and response.

**Comment by jeromelaban** [comment]
@lindexi which platforms are you targeting? You may be able to use the Composition API, or you can use SkiaSharp directly. Which features from Microsoft.Maui.Graphics are of interest to you?

**Reply by lindexi** [reply]
@jeromelaban Thank you very much. Can I find the document about use the SkiaSharp directly? Such as drawing something to the control.

**Reply by lindexi** [reply]
> which platforms are you targeting?

The WPF and GTK.

**Reply by jeromelaban** [reply]
When targeting WPF and GTK, you can use either the WinUI Composition API or SkiaSharp.

**Reply by lindexi** [reply]
@jeromelaban Suppose I have a FrameworkElement object and I want it to draw my custom ink strokes based on the touch information it receives through Pointer series events. How can I get started?

**Reply by lindexi** [reply]
The biggest issue I'm currently facing is that I'm unsure how to associate SkiaSharp with UNO's FrameworkElement. For instance, I want this FrameworkElement to display the content I've drawn through SkiaSharp. It seems that apart from the WriteableBitmap approach, I haven't found any other feasible methods yet.


```csharp
        private ImageSource RenderBySkia()
        {
            var info = new SKImageInfo(150, 150);
            using var surface = SKSurface.Create(info);
            var canvas = surface.Canvas;
            DoMyRenderCode(canvas); // The business code
            using var image = surface.Snapshot();
            var skBitmap = SKBitmap.FromImage(image);
            var bitmap = new WriteableBitmap(skBitmap.Width, skBitmap.Height);
            using var stream = bitmap.PixelBuffer.AsStream();
            stream.Write(skBitmap.Bytes, 0, skBitmap.Bytes.Length);
            return bitmap;
        }
```

**Reply by jeromelaban** [reply, marked as answer]
This sample works for all uno platforms: https://github.com/unoplatform/Uno.Samples/tree/master/UI/SkiaSharpTest.

You can use a `SKXamlCanvas` like any other control to draw Skia content.

**Reply by lindexi** [reply]
@jeromelaban Thank you very much. And I will try the demo.

**Reply by lindexi** [reply]
@jeromelaban Sorry, I run the demo but It didn't work as expected.

At first, I find the running error as `If you are seeing this, make sure to follow the "Migrating WpfHost" section of Migrating from previous releases article in the Uno Platform documentation at https://aka.platform.uno/uno5-wpfhost-migration. WpfHost is used at the application level instead of window level starting Uno Platform 5.0.`. And Then I fixed it in https://github.com/unoplatform/Uno.Samples/pull/627.

And now, it will throw the System.NotSupportedException: `SKSwapChainPanel is not supported for Skia based platforms` when I run the SkiaSharpTest.Skia.WPF project.

**Reply by lindexi** [reply]
Emmmm...

After I remove the `SKSwapChainPanel` from MainPage.xaml, I can run the SkiaSharpTest.Skia.WPF project without exception. But I find it have to use SKXamlCanvas to render.

The SKXamlCanvas is working the hack thing. The SKXamlCanvas will render and then copy to WriteableBitmap to show... I do not think the SKXamlCanvas do the right thing.

**Reply by lindexi** [reply]
How about public the `Visual.Render` method?

https://github.com/unoplatform/uno/blob/44c7d265402ded55ff077ff003d19249a6a8ea2b/src/Uno.UI.Composition/Composition/Visual.skia.cs#L85

**Reply by lindexi** [reply]
@jeromelaban Thank you for your attention. Here is my new progress:

I've written a demo code for a hack, designed to use the `Draw` method of Visual in Uno.Skia applications for interface rendering. My approach was to create a project named `SamplesApp`. This is because UNO has written the InternalsVisibleToAttribute feature for this AssemblyName, which allows access to internal members within this project, making it easier for me to write hack code.

I've created a type named GraphicsCanvasVisual that inherits from Visual, allowing this type to override the Draw method. This means I can use GraphicsCanvasVisual to participate in the rendering of the VisualTree interface tree. I then wrote another type named GraphicsCanvasElement that inherits from FrameworkElement. In the constructor of this type, I added GraphicsCanvasVisual to the `Visual.Children` of GraphicsCanvasElement.

By simply adding GraphicsCanvasElement to XAML, I can render the content I drew in GraphicsCanvasVisual on the interface. I feel that I can encapsulate support for Microsoft.Maui.Graphics based on this. 

You can find my demo in https://github.com/lindexi/lindexi_gd/tree/5db72c3d67f121becefb5c4e8a0eacd5099c4198/KefalurcilaybelJallbuderenajel

**Reply by lindexi** [reply]
@jeromelaban Thank you for your attention. Here is my new progress:

I have been attempting some API design and implementation. Now, I am able to interface with Microsoft.Maui.Graphics using the following code in XAML. I tested this using my internal code and found that using Microsoft.Maui.Graphics in conjunction with overriding the Visual's Draw method is at least 3 times faster than using SKXamlCanvas.

```xml
xmlns:graphics="using:Microsoft.Maui.Graphics.UnoAbstract"

<graphics:GraphicsCanvas Draw="GraphicsCanvas_OnDraw"/>
```

The `GraphicsCanvas_OnDraw` method is:

```csharp
using Microsoft.Maui.Graphics;

    private void GraphicsCanvas_OnDraw(object? sender, Microsoft.Maui.Graphics.ICanvas e)
    {
        e.StrokeSize = 5;
        e.StrokeColor = Colors.Red;
        e.DrawRectangle(0, 0, 100, 100);
    }
```

You can find my demo in https://github.com/lindexi/lindexi_gd/tree/daf3e44a853177d55e9ebb15d989e27b1e497591/KefalurcilaybelJallbuderenajel

![](https://github.com/unoplatform/uno/assets/16054566/6034c96e-f93f-4b94-978a-869124a1c2be)

**Reply by jeromelaban** [reply]
I'm not sure to understand why SKXamlCanvas does not fit what you need. What does Maui Graphics bring that Skia does not?

**Reply by lindexi** [reply]
Because the SKXamlCanvas is working the hack thing. The SKXamlCanvas will render and then copy to WriteableBitmap to show. 

See https://github.com/mono/SkiaSharp/blob/9274aeec807fd17eec2a3266ad4c2475c37d8a0c/source/SkiaSharp.Views.Uno/SkiaSharp.Views.Uno.WinUI.Skia/SKXamlCanvas.Skia.cs#L27

The SKXamlCanvas have to copy the pixel data to WriteableBitmap :

```csharp
	public partial class SKXamlCanvas : Canvas
	{
		private void DoInvalidate()
		{
			...

			var info = CreateBitmap(out var unscaledSize, out var dpi);

			using (var surface = SKSurface.Create(info, pixelsHandle.AddrOfPinnedObject(), info.RowBytes))
			{
				...

				OnPaintSurface(...);
			}

			// This implementation is not fast enough, and providing the original pixel buffer
			// is needed, yet the internal `IBufferByteAccess` interface is not yet available in Uno.
			// Once it is, we can replace this implementation and provide the pinned array directly
			// to skia.
			using (var data = bitmap.PixelBuffer.AsStream())
			{
				data.Write(pixels, 0, pixels.Length);
				data.Flush();
			}

			bitmap.Invalidate();
		}
        }
```

As stated in the code comments, "This implementation is not fast enough". Upon testing my internal code, I found that SKXamlCanvas is indeed slow, at least three times slower than drawing from Visual's Draw. For many business requirements that demand even a slight rendering performance, using SKXamlCanvas is far from meeting the minimum standard, such as dynamic chart displays and handwriting drawing. Especially in the business requirements of handwriting drawing, drawing latency and drawing performance are extremely important. A lower rendering latency can bring significant competitiveness to handwriting applications and whiteboard applications.

**Reply by lindexi** [reply]
The Maui Graphics is the API define which not limit in skia. And the Maui Graphics can help me to re-use the exists code.

**Reply by lindexi** [reply]
@jeromelaban Consider introducing controls similar to the CanvasControl type in Win2d of UWP, or the [GraphicsView](https://learn.microsoft.com/en-us/dotnet/maui/user-interface/graphics/?view=net-maui-8.0) type in MAUI, or DrawingContext in WPF. I anticipate the introduction of such controls in UNO, which would allow developers direct access to basic, low-level drawing. These controls would enable developers to have better, more detailed control over their drawing content. With these controls, developers could better implement customized, complex interface drawing logic. Furthermore, many existing infrastructures built on basic graphics can continue to be established on top of this. I believe this would greatly benefit the UNO development ecosystem.

**Reply by jeromelaban** [reply]
We already have SkiaSharp and Composition API for this purpose. (even if the current SkiaSharp implementation uses an intermediate surface to render). It's great that you were able to use you own tooling.

**Reply by lindexi** [reply]
May I kindly inquire if it would be permissible for me to create a Package that would provide a more direct interface into the rendering layer of UNO? In other words, would it be acceptable to create a Package using the technology we previously discussed? As I explained earlier, I am not interested in building a rendering framework on top of SkiaSharp's SKXamlCanvas. The reason being, SKXamlCanvas seems to be more of a hacking solution and its performance is subpar, failing to meet the minimum performance requirements.

Additionally, I am contemplating modifying the SKXamlCanvas type in SkiaSharp. When UNO uses Skia, such as Gtk.Skia or Wpf.Skia, I would like SKXamlCanvas to directly utilize UNO's Skia surface, instead of creating an additional surface and copying the rendering data to a WriteableBitmap via the CPU. I am unsure if such optimization of SKXamlCanvas is encouraged or not.

Thank you @jeromelaban

**Reply by jeromelaban** [reply]
We have the intent of providing a way to create such interoperability with SkiaSharp, but at this time, there's no defined API that allows to do so. The main problem here is the type of API that will be needed to do so, and it will need to be at the composition level to avoid threading issues, similar to what the [`TextVisual`](https://github.com/unoplatform/uno/blob/6fe692ccbff5a7ba25f2e0c9ca85002daad8cc6d/src/Uno.UI/UI/Xaml/Controls/TextBlock/TextVisual.skia.cs#L21) is doing, for instance.

Also, replicating the behavior of the SKXamlCanvas completely is required, and it must behave like an independent surface as it currently does, let's keep track here: https://github.com/unoplatform/uno/issues/15166.

**Reply by lindexi** [reply]
Thank you @jeromelaban 

Before the SkiaVisual added, should I continue use my hack code？

**Reply by jeromelaban** [reply]
A hack is a hack. If you rely on private APIs, your code may be broken with any Uno release.

**Reply by lindexi** [reply]
@jeromelaban Thank you for your support, I believe I have grasped your point. I plan to write a blog post documenting our above-mentioned technical discussion.

**Reply by lindexi** [reply]
I wrote the blog about this here: https://blog.lindexi.com/post/How-to-Integrate-Microsoft.Maui.Graphics-into-the-UNO-Framework-in-dotnet.html

**Reply by lindexi** [reply]
And https://medium.com/@lindexi_gd/how-to-integrate-microsoft-maui-graphics-into-the-uno-framework-in-dotnet-cd9705d96572
